pkg/genericapiserver/storage_factory.go

```diff
@@ -217,3 +233,39 @@ func (s *DefaultStorageFactory) Backends() []string {
 	}
 	return backends.List()
 }
+
+// NewStorageCodec assembles a storage codec for the provided storage media type, the provided serializer, and the requested
+// storage and memory versions.
+func NewStorageCodec(storageMediaType string, ns runtime.StorageSerializer, storageVersion, memoryVersion unversioned.GroupVersion, config storagebackend.Config) (runtime.Codec, error) {
+	mediaType, options, err := mime.ParseMediaType(storageMediaType)
+	if err != nil {
+		return nil, fmt.Errorf("%q is not a valid mime-type", storageMediaType)
+	}
+	serializer, ok := ns.SerializerForMediaType(mediaType, options)
+	if !ok {
+		return nil, fmt.Errorf("unable to find serializer for %q", storageMediaType)
+	}
+
+	s := serializer.Serializer
+
+	// etcd2 only supports string data - we must wrap any result before returning
```

Comment:
for etcd2 yes. etcd2 only accepts json. json string cannot contain random bytes, you have to base64 it.